### PR TITLE
Add zeroing function for arrays

### LINKED ISSUE #1675: initializing empty arrays
since https://github.com/ethereum/vyper/issues/1493 arrays can't be left uninitialized. we should introduce a way to create a freshly zeroed array, along the lines of `zero_array(N)`, or maybe have a special case for arrays to be uninitialized (that does explicitly set them to zero).

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
- Fix #1675 by introducing syntax for initializing new arrays with zero

### How I did it
- Get rid of NullType from codebase. Zeroing operations are now from the None sentinel value alone.
- Factor mzero into a function

### How to verify it
See tests

### Description for the changelog
- Introduce syntax for initializing new arrays with zero
- Check array assignment is always between arrays of same length

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/parser/parser_utils.py [hunk 1]
```diff
@@ -889,11 +889,18 @@ def annotate_ast(
     RewriteUnarySubVisitor().visit(parsed_ast)
 
 
+def mzero(dst, nbytes):
```
Comment (jacqueswww) [inline vyper/parser/parser_utils.py]: Yes! This is great much clearer intent :-)

**Comment by fubuloubu** [comment]
Can we use `empty` instead of `clear`? I don't like mixing the semantics of the two. Let's keep `clear` as clearing/resetting storage space, and `empty` as creating empty memory.

**Comment by fubuloubu** [comment]
can you rebase on master?

**Comment by charles-cooper** [comment]
> can you rebase on master?

@fubuloubu is this because you force pushed to master :joy: ? in the future can we protect master from all pushes except from PRs?

**Comment by fubuloubu** [comment]
Maybe, that wasn't the commit that was lost though

File: vyper/parser/lll_node.py [hunk 2]
```diff
@@ -213,7 +212,8 @@ def __init__(self,
                     self.gas = sum([arg.gas for arg in self.args]) + 30
                 if self.value == 'if_unchecked':
                     self.gas = self.args[0].gas + self.args[1].gas + 17
-        elif self.value is None and isinstance(self.typ, NullType):
+        elif self.value is None:
+            # TODO revisit this
```
Comment (fubuloubu) [inline vyper/parser/lll_node.py]: Needs a way better comment than this if you're going to leave a TODO

Comment (charles-cooper) [reply vyper/parser/lll_node.py]: thanks, resolved

**Comment by charles-cooper** [comment]
some work on the gas estimator is still needed

**Comment by fubuloubu** [comment]
Rebase on master to get updated CI

**Comment by charles-cooper** [comment]
> some work on the gas estimator is still needed

I fixed it.

**Comment by fubuloubu** [comment]
Meeting notes: Change to `empty`, leave current use of `clear` for now, but could be deprecated in the future.

**Comment by michwill** [comment]
What would be pythonic is something like `a: uint256[10] = [0 for _ in range(10)]`, or even `a: uint256[10] = [0] * 10`

**Comment by fubuloubu** [comment]
@michwill the later would be the better choice, but we had a very long conversation about this and enabling the Pythonic list comprehension syntax is a rabbit hole, plus `empty` is more broadly useful.

File: vyper/functions/functions.py [hunk 3]
```diff
@@ -1225,12 +1225,11 @@ def sqrt(expr, args, kwargs, context):
...
 
-def _clear():
-    raise CompilerPanic(
-        "This function should never be called! `clear()` is currently handled "
-        "differently than other functions as it self modifies its input argument "
-        "statement. Please see `_clear()` in `stmt.py`"
-    )
+def empty(expr, context):
+    if len(expr.args) != 1:
+        raise ParserException('function expects two parameters.', expr)
```
Comment (fubuloubu) [inline vyper/functions/functions.py:1230]:
```suggestion
        raise CompilerPanic('function expects one parameter.', expr)
```

File: vyper/parser/parser_utils.py [hunk 4]
```diff
@@ -12,6 +12,7 @@
     InvalidLiteral,
     StructureException,
     TypeMismatch,
+    CompilerPanic,
```
Comment (fubuloubu) [inline vyper/parser/parser_utils.py:15]: Put this import in alphabetical order

**Comment by charles-cooper** [comment]
I'm OK merging this with the caveat that users might have some trouble passing empty to and from functions. The compiler will probably just complain. Should be fixed once arg packing / unpacking is refactored.

File: vyper/functions/functions.py (hunk 3 above)
Comment (iamdefinitelyahuman) [reply vyper/functions/functions.py:1230]: I think `ArgumentException` is the most fitting exception to raise here.  `ParserException` implies that the code can't be parsed. It also doesn't support annotation.

File: vyper/parser/expr.py [hunk 5]
```diff
@@ -257,7 +256,10 @@ def constants(self):
                 pos=getpos(self.expr),
             )
         elif self.expr.value is None:
-            return LLLnode.from_list(None, typ=NullType(), pos=getpos(self.expr))
+            # block None
+            raise InvalidLiteral(
+                    'None is not allowed in vyper'
+                    '(use a default value or built-in `clear()`')
```
Comment (iamdefinitelyahuman) [inline vyper/parser/expr.py:262]:
```suggestion
                    '(use a default value or built-in `empty()`')
```

File: vyper/parser/parser_utils.py [hunk 6]
```diff
@@ -156,16 +174,17 @@ def make_byte_slice_copier(destination, source, length, max_length, pos=None):
     elif destination.location == "storage":
         setter = ['sstore', ['add', '_opos', ['mload', MemoryPositions.FREE_LOOP_INDEX]], loader]
     else:
-        raise Exception("Unsupported location:" + destination.location)
+        raise CompilerPanic("Unsupported location:" + destination.location)
```
Comment (iamdefinitelyahuman) [inline vyper/parser/parser_utils.py:177]:
```suggestion
        raise CompilerPanic(f"Unsupported location: {destination.location}")
```

File: vyper/parser/stmt.py [hunk 7]
```diff
@@ -120,11 +124,15 @@ def _check_valid_assign(self, sub):
                     raise TypeMismatch('Invalid type, expected: bytes32', self.stmt)
                 return
             else:
-                raise TypeMismatch('Invalid type, expected: bytes32', self.stmt)
+                raise TypeMismatch(
+                    f"Invalid type, expected: bytes32", self.stmt
```
Comment (iamdefinitelyahuman) [inline vyper/parser/stmt.py:128]:
```suggestion
                    "Invalid type, expected: bytes32", self.stmt
```

File: vyper/functions/functions.py [hunk 8]
```diff
@@ -7,9 +7,9 @@
     ast as vy_ast,
 )
 from vyper.exceptions import (
-    CompilerPanic,
     ConstancyViolation,
```
Comment (fubuloubu) [inline vyper/functions/functions.py:10]:
```suggestion
    ArgumentException,
    ConstancyViolation,
```

File: vyper/functions/functions.py [hunk 9]
```diff
@@ -7,9 +7,9 @@
...
     InvalidLiteral,
+    ParserException,
```
Comment (fubuloubu) [inline vyper/functions/functions.py:12]:
```suggestion
```

**Comment by or2008** [comment]
hello, how can I use empty to mimic solidity's `new bytes(0)`?

**Comment by fubuloubu** [comment]
@or2008 can you do us a favor and create a new issue with the source code that is giving you the problem? Thanks!

**Comment by or2008** [comment]
sure, opened https://github.com/vyperlang/vyper/issues/2112

**Pull request merged**